src/Domain/Creators/BinaryCompactObject.cpp

```diff
@@ -655,17 +646,27 @@ Domain<3> BinaryCompactObject::create_domain() const {
 
 std::unordered_map<std::string,
                    std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>
-BinaryCompactObject::functions_of_time() const {
+BinaryCompactObject::functions_of_time(
+    const std::unordered_map<std::string, double>& initial_expiration_times)
+    const {
   std::unordered_map<std::string,
                      std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>
       result{};
   if (not enable_time_dependence_) {
     return result;
   }
 
-  const double initial_expiration_time =
-      initial_expiration_delta_t_ ? initial_time_ + *initial_expiration_delta_t_
-                                  : std::numeric_limits<double>::infinity();
+  if (not initial_expiration_times.empty()) {
```

Comment:
If I understand correctly, the control flow goes like this: The domain creator decides which FoT are around to control the domain. Then control systems can jump in to update those FoT. So it seems to me the domain creator doesn't need to know anything about control systems, but only the reverse: Control systems are set up with a name, and that name must exist as a FoT in the domain. Therefore, it seems to me the initial expiration time shouldn't be set by the domain creator, but the control system should jump in after the FoT were created to set the expiration time (essentially saying "I'm going to take control of this FoT from now on, here's when I'm going to update it"). I think my suggestion is to set the initial expiration time where you initialize the control systems. Indeed I think you are currently in danger of running into a cyclic dependency: Control systems will depend on Domain/TimeDep in some way or another (no way around that, right?), so Domain/TimeDep can't depend on ControlSystems. Please make sure you are keeping this dependency graph in mind.

Reply:
Good point, but I'm not sure how to deal with it.  It might make sense to move most of this logic from the domain creators to the FoT initialization code, but that's still in Domain so that doesn't help.  We can't run initialization actions on the global cache.  We can't handle it entirely from the control system chares, because something has to handle the FoTs that aren't being controlled.

Reply:
Ok. My suggestions for now are:

- Don't include ControlSystem things in `Domain/` unless you absolutely have to for now.
- Remove the `initial_expiration_times` argument from the domain creators. Instead, set it after retrieving the FoT from the domain creator. That's currently in `Domain/FoT/Tags.hpp` and I suppose it's ok to leave it there for now, but we'll have to think where we can move it to avoid a cyclic dependency.

Reply:
My issue with this, then, is what do you set the initial expiration times as in the DomainCreator? Because that's where the functions of time are constructed with all their initial data. Functions of time need to be constructed with an initial expiration time as well.

I could conceivably set the "expiration time" in the DomainCreator to the initial time as a sort of fake expiration time, then reset the initial expiration time to the actual initial expiration time in `Domain/FoT/Tags/hpp`. However, I don't like resetting an expiration time without any update actually happening.

Reply:
Can the initial expiration time be max, inf, nan or std::nullopt, signaling the FoT is "uncontrolled", and then the control system takes over and gives it an expiration time? Then there can also be a check somewhere that all FoTs that require a control system have one.

Reply:
Currently, expiration times must be increasing from call to call. The initial expiration time could be -inf or -max, then when the control system sets the actual expiration times, if there is a function not being controlled, its expiration time could be set to +inf then.

Reply:
Maybe we could have each type (C++ sense) of FoT either always be controlled or always be uncontrolled.  An uncontrolled piecewise polynomial always struck me as somewhat questionable, except for the linear case.  Then, I think the setup could be done by the control components during initialization.  (Although I don't know where a sanity check for missing control systems would go.)